### PR TITLE
Tags filter support

### PR DESCRIPTION
**Pull request by RichardChen820**
**Tags filter support:**

* Added a `Tags` field to the `AddSettingOptions` and `SetSettingOptions` structs, enabling users to assign key-value metadata tags when creating or updating configuration settings. 
* Updated the `AddSetting` and `SetSetting` methods in `client.go` to include the new `Tags` field when constructing a `Setting`. 
* Modified the `SettingSelector` struct to support a `TagsFilter` field, allowing retrieval of settings filtered by tags. The filter is passed to the generated API calls.

**Documentation and examples:**

* Updated the README to describe how tags can be used for metadata and filtering, with examples of tag-based queries.
* Added a new example in `examples_test.go` demonstrating how to create a setting with tags and retrieve settings using a tag filter.

**REST API version update:**

* Bumped the Azure App Configuration REST API version from `2023-10-01` to `2023-11-01` in `autorest.md` and `appconfiguration_ext.json` to enable tag features.

### CODE REVIEW

File: sdk/data/azappconfig/options.go [hunk 1]
```diff
@@ -19,6 +19,10 @@ type AddSettingOptions struct {
 
 	// Configuration setting label.
 	Label *string
+
+	// A dictionary of tags used to assign additional properties to a configuration setting.
+	// These can be used to indicate how a configuration setting may be applied.
+	Tags map[string]string
```
Comment (linglingye001) [inline sdk/data/azappconfig/options.go]: why using map[string]string in `AddSettingsOptions`? Is it by design?

Comment (RichardChen820) [reply sdk/data/azappconfig/options.go]: This should be a designed API, it should've tags support at the time of adding `AddSetting()` API, we need to confirm with the initial author about the design of it (asking them to review this PR is good enough I think).

Here. I just speculate the initial intention and propose my fix to this bug, two reasons:
1. It initializes the [`setting`](https://github.com/Azure/azure-sdk-for-go/blob/038f9dc38aaad217983ddbefe5314acfe888bc0b/sdk/data/azappconfig/client.go#L101C1-L101C101) and pass it to `PutKeyValue` function, and tags in `Setting` type is map[string]string , we just keep consistency is fine
2. It comply with other languages e.g., [dotnet](https://github.com/Azure/azure-sdk-for-net/blob/50485990713dd742ce3127ceff833bed2eda4c87/sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient.cs#L279C8-L308C10), `Tags` is also a dictionary property of `ConfigurationSetting` while adding the setting. dotnet has a slightly different API design becuase it has method overload, but `Go` doesn't, so I think that's why the caller needs to pass `Tags` in `AddSettingsOptions`. But its type is clear, it is supposed to be a dictionary in go as well.

File: sdk/data/azappconfig/setting.go [hunk 2]
```diff
@@ -78,7 +78,8 @@ func toGeneratedETagString(etag *azcore.ETag) *string {
 func (cs Setting) toGenerated() generated.KeyValue {
 	tags := make(map[string]*string)
```
Comment (linglingye001) [inline sdk/data/azappconfig/setting.go]: In generated.KeyValue, the tags is `map[string]*string`, and in `Setting` we have `map[string]string`, do you know why?

Comment (RichardChen820) [reply sdk/data/azappconfig/setting.go]: Have no idea about the initial reason for doing so, may need input from @jhendrixMSFT @RickWinter

**Review by zhiyuanliang-ms — APPROVED** [review]

**Comment by RickWinter** [comment]
@jhendrixMSFT Can you review. Also since the appconfig team now owns this SDK we should add the appropriate folks to the CODEOWNER file.

File: sdk/data/azappconfig/setting.go (hunk 2 above)
Comment (jhendrixMSFT) [reply sdk/data/azappconfig/setting.go]: The generator emits as `map[string]*type` (where `type` isn't implicitly nil-able) for a few reasons.
- this is for compat when switching to/from a dictionary with nullable values (we don't want such changes to be breaking)
- some services support sending JSON `null` for dictionary values to mean "delete this entry"

If you know your service will never need the map's value type to be nil-able, then exposing it as `map[string]string` is the right call as it has better ergonomics.

File: sdk/data/azappconfig/options.go (hunk 1 above)
Comment (jhendrixMSFT) [reply sdk/data/azappconfig/options.go]: All correct. A method's options-type (i.e. the last param where the type's suffix is `Options`) is an extensibility point to add optional method params due to lack of overloads.

Note that for fields in these types, the zero-value _must not_ be ambiguous.  i.e. the zero-value means "don't send anything for this field" which is why a lot of fields are pointer-to-type.

Since maps are nil-able, what you have makes sense.

**Comment by jhendrixMSFT** [comment]
@RichardChen820 can you please help us get the list of appropriate aliases for us to update the `CODEOWNERS` file?

File: sdk/data/azappconfig/README.md
Comment (jhendrixMSFT) [inline sdk/data/azappconfig/README.md:1]: Please also update the `CHANGELOG.md` file noting the new features.

In addition, the version must also be updated to `1.3.0` in the changelog and [here](https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/data/azappconfig/version.go#L11).

Do you think this should go out as an initial beta? If so, then keep the `-beta.1` suffix.

**Comment by RichardChen820** [comment]
> @RichardChen820 can you please help us get the list of appropriate aliases for us to update the `CODEOWNERS` file?

Please add me and Lingling: @RichardChen820 @linglingye001

File: sdk/data/azappconfig/README.md
Comment (RichardChen820) [reply sdk/data/azappconfig/README.md:1]: I think we can just go with 1.3.0 stable. Is the 1.2.1-beta.1 planned but hasn't been released yet?

File: sdk/data/azappconfig/setting.go (hunk 2 above)
Comment (RichardChen820) [reply sdk/data/azappconfig/setting.go]: I checked our dataplane backend, and we don’t have any restriction on non-null tag values on the server side, null value is allowed. I also reviewed the[ .NET SDK implementation](https://github.com/Azure/azure-sdk-for-net/blob/c0e3f8082be4812792466d68e068587370fe891f/sdk/appconfiguration/Azure.Data.AppConfiguration/src/Models/ConfigurationSetting.cs#L26C1-L26C9), and it doesn’t prohibit null values either. So I’m leaning toward using `map[string]*string` for `Tags` in `Setting`. However, this would be a breaking change. Any concerns? @jhendrixMSFT

Comment (jhendrixMSFT) [reply sdk/data/azappconfig/setting.go]: I don't think we should make such a change unless there's loss of functionality and a demand for it to be fixed. Does the exclusion of `nil` for a value preclude any scenarios from working?

File: sdk/data/azappconfig/README.md
Comment (jhendrixMSFT) [reply sdk/data/azappconfig/README.md:1]: No. After `v1.2.0` was released, the version is automatically bumped to the next patch-beta version as a placeholder.

**Comment by RickWinter** [comment]
> > @RichardChen820 can you please help us get the list of appropriate aliases for us to update the `CODEOWNERS` file?
> 
> Please add me and Lingling: @RichardChen820 @linglingye001

You can edit the CODEOWNERS file and submit as part of this PR or you can make a separate PR. I'll approve

**Comment by RichardChen820** [comment]
> You can edit the CODEOWNERS file and submit as part of this PR or you can make a separate PR. I'll approve

@RickWinter please approve this add owner PR https://github.com/Azure/azure-sdk-for-go/pull/25313

File: sdk/data/azappconfig/setting.go (hunk 2 above)
Comment (RichardChen820) [reply sdk/data/azappconfig/setting.go]: This does introduce some functionality issues. As I mentioned, since our dataplane backend allows users to assign a `null pointer` value to a tag, under the current implementation the tag would be dropped when the user retrieves the key-value. This is clearly problematic because an item that should exist gets lost.

If we instead treat `null` as an empty string, that leads to another issue: when the user writes the value back, its original value would be unintentionally altered (from `null` to empty string).

Comment (RichardChen820) [reply sdk/data/azappconfig/setting.go]: So I’d prefer to make the change now, rather than waiting until users start complaining about it.

Comment (jhendrixMSFT) [reply sdk/data/azappconfig/setting.go]: Got it, thanks for clarifying.

I still hesitate to fix this though as this has been like this for almost two years and nobody complained about it. Plus, introducing a new major version does come with some customer cost (and some cost for us, like updating docs/examples etc).

That said, if you think the cost is justified, I'll go along with it.

Ok after writing the above, I was going through the issues, and I found [this one](https://github.com/Azure/azure-sdk-for-go/issues/22303). Is this due to what we're discussing?

Comment (RichardChen820) [reply sdk/data/azappconfig/setting.go]: That’s I think why we should make the breaking change for doing the right thing now, while there are still few users using Tags in Go,

Comment (RichardChen820) [reply sdk/data/azappconfig/setting.go]: [This one ](https://github.com/Azure/azure-sdk-for-go/issues/22303) is another problem, I think we should add a variable to indicate the "Null" value of label to stop confusing users.

File: sdk/data/azappconfig/version.go [hunk 3]
```diff
@@ -8,5 +8,5 @@ package azappconfig
 
 const (
 	moduleName    = "github.com/Azure/azure-sdk-for-go/sdk/data/azappconfig"
-	moduleVersion = "v1.2.1-beta.1"
+	moduleVersion = "v1.3.0"
```
Comment (jhendrixMSFT) [inline sdk/data/azappconfig/version.go]: Breaking changes must increment the major version number, so this would become `v2.0.0`.  In addition, you'll need to perform the following changes.
- update the `go.mod` file to add a `/v2` major version suffix to the module's identity
- update all `import` statements to reference `v2`
- update all docs and examples to use `v2`

Comment (RichardChen820) [reply sdk/data/azappconfig/version.go]: @jhendrixMSFT Updated, please help take a look to see if I missed anything.

File: sdk/data/azappconfig/version.go [hunk 4]
```diff
@@ -7,6 +7,6 @@
 package azappconfig
 
 const (
-	moduleName    = "github.com/Azure/azure-sdk-for-go/sdk/data/azappconfig"
-	moduleVersion = "v1.2.1-beta.1"
+	moduleName    = "github.com/Azure/azure-sdk-for-go/sdk/data/azappconfig/v2"
```
Comment (jhendrixMSFT) [inline sdk/data/azappconfig/version.go]: The name needs to omit the major version suffix.

File: sdk/data/azappconfig/version.go (hunk 3 above)
Comment (jhendrixMSFT) [reply sdk/data/azappconfig/version.go]: The value for `moduleName` must not include the major version suffix. Everything else looks good.

File: sdk/data/azappconfig/version.go (hunk 4 above)
Comment (RichardChen820) [reply sdk/data/azappconfig/version.go]: Ah, right, updated.

File: sdk/data/azappconfig/client_test.go [hunk 5]
```diff
@@ -767,3 +767,478 @@ func CleanupSnapshot(client *azappconfig.Client, snapshotName string) error {
...
+	)
+
+	client := NewClientFromConnectionString(t)
+
+	// Test tags with special characters
+	tags := map[string]*string{
+		"app-name":    to.Ptr("my-app"),
+		"team.owner":  to.Ptr("backend-team"),
+		"cost_center": to.Ptr("engineering"),
+		"env_type":    to.Ptr("staging-test"),
```
Comment (avanigupta) [inline sdk/data/azappconfig/client_test.go]: Can we use some other special chars? `.`, `-` and `_` are not encoded in the URL if that's the special char handling you are trying to verify.

File: sdk/data/azappconfig/client_test.go [hunk 6]
```diff
@@ -767,3 +767,478 @@ func CleanupSnapshot(client *azappconfig.Client, snapshotName string) error {
...
+	require.True(t, found, "Setting with special character tags not found")
+
+	// Clean up
+	_, err = client.DeleteSetting(context.Background(), key, &azappconfig.DeleteSettingOptions{
+		Label: to.Ptr("special-char-test"),
+	})
+	require.NoError(t, err)
+}
+
+func TestEmptyTagsHandling(t *testing.T) {
```
Comment (avanigupta) [inline sdk/data/azappconfig/client_test.go:1027]: Could you add a test for filtering by tags with empty or null value.
Eg: 
- `/kv?tags=EmptyTagVal=&api-version=...`
- `/kv?tags=NullTagVal=%00&api-version=...`

File: sdk/data/azappconfig/internal/generated/zz_azureappconfiguration_client.go [hunk 7]
```diff
@@ -148,6 +148,11 @@ func (client *AzureAppConfigurationClient) checkKeyValuesCreateRequest(ctx conte
 	if options != nil && options.Snapshot != nil {
 		reqQP.Set("snapshot", *options.Snapshot)
 	}
+	if options != nil && options.Tags != nil {
+		for _, qv := range options.Tags {
+			reqQP.Add("tags", qv)
```
Comment (avanigupta) [inline sdk/data/azappconfig/internal/generated/zz_azureappconfiguration_client.go:153]: This looks like autogenerated code, but just to confirm - this will generate multiple occurrences of `tags` query parameter in the URL if there are multiple tag filters, right?
Eg: `/kv?tags={filter1}&tags={filter2}&tags={filter3}`

File: sdk/data/azappconfig/internal/generated/zz_options.go [hunk 8]
```diff
@@ -49,17 +49,20 @@ type AzureAppConfigurationClientCheckKeyValuesOptions struct {
...
 	Key *string
 
-	// A filter used to match labels
+	// A filter used to match labels. Syntax reference: https://aka.ms/azconfig/docs/keyvaluefiltering
 	Label *string
 
 	// Used to select what fields are present in the returned resource(s).
 	Select []SettingFields
 
 	// A filter used get key-values for a snapshot. Not valid when used with 'key' and 'label' filters.
```
Comment (avanigupta) [inline sdk/data/azappconfig/internal/generated/zz_options.go]:
```suggestion
	// A filter used get key-values for a snapshot. Not valid when used with 'key', 'label' or `tags` filters.
```

File: sdk/data/azappconfig/internal/generated/zz_options.go [hunk 9]
```diff
@@ -196,6 +202,9 @@ type AzureAppConfigurationClientGetKeyValuesOptions struct {
 	// A filter used get key-values for a snapshot. The value should be the name of the snapshot. Not valid when used with 'key'
 	// and 'label' filters.
```
Comment (avanigupta) [inline sdk/data/azappconfig/internal/generated/zz_options.go]:
```suggestion
	// A filter used get key-values for a snapshot. The value should be the name of the snapshot. Not valid when used with 'key',
	// 'label' or `tags` filters.
```

File: sdk/data/azappconfig/options.go [hunk 10]
```diff
@@ -19,6 +19,10 @@ type AddSettingOptions struct {
 
 	// Configuration setting label.
 	Label *string
+
+	// A dictionary of tags used to assign additional properties to a configuration setting.
+	// These can be used to indicate how a configuration setting may be applied.
```
Comment (avanigupta) [inline sdk/data/azappconfig/options.go]: Not sure I follow the second line.
Can we just say "Configuration setting tags"? Or maybe keep only the first line of your summary?

File: sdk/data/azappconfig/setting_selector.go [hunk 11]
```diff
@@ -30,6 +30,11 @@ type SettingSelector struct {
 	// Label filter that will be used to select a set of configuration setting entities.
 	LabelFilter *string
 
+	// Tags filter that will be used to select a set of configuration setting entities.
+	// This is a list of comma-separated tags. For more information about filtering by tags, see:
+	// https://learn.microsoft.com/azure/azure-app-configuration/rest-api-key-value?pivots=v23-11#filtering
```
Comment (avanigupta) [inline sdk/data/azappconfig/setting_selector.go]: Can use short link here

File: sdk/data/azappconfig/setting_selector.go [hunk 12]
```diff
@@ -30,6 +30,11 @@ type SettingSelector struct {
 	// Label filter that will be used to select a set of configuration setting entities.
 	LabelFilter *string
 
+	// Tags filter that will be used to select a set of configuration setting entities.
+	// This is a list of comma-separated tags. For more information about filtering by tags, see:
```
Comment (avanigupta) [inline sdk/data/azappconfig/setting_selector.go]:
```suggestion
	// This is a list of tag filters in the format {tagName=tagValue}. For more information about filtering by tags, see:
```

File: sdk/data/azappconfig/options.go (hunk 10 above)
Comment (RichardChen820) [reply sdk/data/azappconfig/options.go]: Make sense, let's only keep this first line.

File: sdk/data/azappconfig/internal/generated/zz_azureappconfiguration_client.go (hunk 7 above)
Comment (RichardChen820) [reply sdk/data/azappconfig/internal/generated/zz_azureappconfiguration_client.go:153]: Right

File: sdk/data/azappconfig/client_test.go (hunk 6 above)
Comment (RichardChen820) [reply sdk/data/azappconfig/client_test.go:1027]: added

File: sdk/data/azappconfig/client_test.go (hunk 5 above)
Comment (RichardChen820) [reply sdk/data/azappconfig/client_test.go]: updated

**Review by linglingye001 — APPROVED** [review]

File: sdk/data/azappconfig/client_test.go [hunk 13]
```diff
@@ -1242,3 +1242,243 @@ func TestTagsFilteringEndToEnd(t *testing.T) {
...
+		require.Contains(t, foundKeys, "setting-with-normal-tags")
+		require.Contains(t, foundKeys, "setting-with-nil-tag-value")
+		require.Contains(t, foundKeys, "setting-with-empty-tag-value")
+		require.NotContains(t, foundKeys, "setting-mixed-tags") // has env=staging
+	})
+
+	// Test 2: Filter by tag key regardless of value (including nil/empty)
+	t.Run("TagKeyExistsFilter", func(t *testing.T) {
+		// Note: This test depends on how the backend handles tag key existence
+		// Some implementations might not return nil-valued tags
```
Comment (avanigupta) [inline sdk/data/azappconfig/client_test.go]: Backend does not support filtering by tag key only (or tag existence). The only supported scenario is filtering by exact match (case sensitive) of tag name and tag value both.

Comment (avanigupta) [reply sdk/data/azappconfig/client_test.go]: The name of the test and comment is misleading, but the test is correctly asserting that only "setting-with-normal-tags" should be returned.

File: sdk/data/azappconfig/client_test.go [hunk 14]
```diff
@@ -1242,3 +1242,243 @@ func TestTagsFilteringEndToEnd(t *testing.T) {
...
+
+			for _, setting := range page.Settings {
+				if strings.HasPrefix(*setting.Key, "setting-") {
+					foundKeys = append(foundKeys, *setting.Key)
+				}
+			}
+		}
+
+		// Should find only the setting with service=api
+		require.Contains(t, foundKeys, "setting-with-normal-tags")
```
Comment (avanigupta) [inline sdk/data/azappconfig/client_test.go:1377]: Could you also do `require.Len(t, foundKeys, 1)` here?

File: sdk/data/azappconfig/setting_selector.go [hunk 15]
```diff
@@ -31,8 +31,8 @@ type SettingSelector struct {
 	LabelFilter *string
 
 	// Tags filter that will be used to select a set of configuration setting entities.
-	// This is a list of comma-separated tags. For more information about filtering by tags, see:
-	// https://learn.microsoft.com/azure/azure-app-configuration/rest-api-key-value?pivots=v23-11#filtering
+	// This is a list of tag filters in the format {tagName=tagValue}. For more information about filtering by tags, see::
```
Comment (avanigupta) [inline sdk/data/azappconfig/setting_selector.go]: Typo - extra colon at the end

File: sdk/data/azappconfig/client_test.go [hunk 16]
```diff
@@ -1242,3 +1242,243 @@ func TestTagsFilteringEndToEnd(t *testing.T) {
...
+		}
+
+		// Should find no settings with non-existent tag
+		require.Empty(t, foundKeys)
+	})
+
+	// Test 7: Filter by tag key that exists with nil values
+	t.Run("TagKeyExistsWithNilValuesFilter", func(t *testing.T) {
+		selector := azappconfig.SettingSelector{
+			TagsFilter: []string{"tag1=\x00"}, // tag1 exists with nil or empty values in some settings
```
Comment (avanigupta) [inline sdk/data/azappconfig/client_test.go]: What happens if you pass "tag1=\0"? I know it's interchangeable since both are URL encoded to "%00", but I'm curious if we prefer one over the other in Go SDK? We normally publicize "\0" as the filter for null value in our docs and sample code.

Comment (RichardChen820) [reply sdk/data/azappconfig/client_test.go]: It's language characteristic of GO, "\0" is not a valid escape sequence in GO.
Go expects either:
  - A recognized single-character escape (\n, \t)
  - A three-digit octal escape (\000)
  - A two-digit hex escape (\x00)
  - A Unicode escape (\u0000)

File: sdk/data/azappconfig/setting_selector.go (hunk 15 above)
Comment (RichardChen820) [reply sdk/data/azappconfig/setting_selector.go]: fixed

File: sdk/data/azappconfig/client_test.go (hunk 13 above)
Comment (RichardChen820) [reply sdk/data/azappconfig/client_test.go]: Good catch, I should've found it, I think it's AI hallucination.

File: sdk/data/azappconfig/client_test.go (hunk 14 above)
Comment (RichardChen820) [reply sdk/data/azappconfig/client_test.go:1377]: added

**Comment by RichardChen820** [comment]
@jhendrixMSFT Does it looks good to you? I'm going to merge it.

File: sdk/data/azappconfig/version.go [hunk 17]
```diff
@@ -8,5 +8,5 @@ package azappconfig
 
 const (
 	moduleName    = "github.com/Azure/azure-sdk-for-go/sdk/data/azappconfig"
-	moduleVersion = "v1.2.1-beta.1"
+	moduleVersion = "v2.0.0"
```
Comment (RickWinter) [inline sdk/data/azappconfig/version.go:11]: Why is this a major version rev when the change is additive? Can this feature be added without a breaking change?

Comment (RichardChen820) [reply sdk/data/azappconfig/version.go:11]: Please refer to the discussion in this thread https://github.com/Azure/azure-sdk-for-go/pull/25194#discussion_r2342843983

Comment (RichardChen820) [reply sdk/data/azappconfig/version.go:11]: To summarize, it still comes down to maintaining feature completeness and consistency across languages. At first, I was hesitant about making a breaking change just to support nil pointer values—it felt a bit excessive. But since it does cause functionality issue, it’s better to fix it now rather than wait until problems arise and customers start complaining.

Moreover, since the SDK itself is quite fundamental, and we have two other projects (https://github.com/Azure/AppConfiguration-KubernetesProvider & https://github.com/Azure/AppConfiguration-GoProvider)  built on top of it, once tag filter support is added, those projects will also inherit the behavior. If we don’t fix this now, the issue will inevitably be exposed to users later on.

**Comment by RickWinter** [comment]
I've cleared my blocking as the change seems reasonable.

NOTE: Per policy the v1 library must be maintained for 12 months after the date of the v2 release before v1 can be deprecated.
